pkg/pushsync/pushsync.go

```diff
@@ -77,59 +78,139 @@ func (ps *PushSync) Close() error {
 	return nil
 }
 
-// handler handles chunk delivery from other node and inserts it to localstore.
-// it also sends this chunk to the closest peer if one exists.
+// handler handles chunk delivery from other node and forwards to its destination node.
+// If the current node is the destination, it stores in the local store and sends a receipt.
 func (ps *PushSync) handler(ctx context.Context, p p2p.Peer, stream p2p.Stream) error {
-	_, r := protobuf.NewWriterAndReader(stream)
+	w, r := protobuf.NewWriterAndReader(stream)
 	defer stream.Close()
 
-	var ch pb.Delivery
+	// Get the delivery
+	ch, err := ps.getChunkDelivery(r)
+	if err != nil {
+		return fmt.Errorf("chunk delivery: %w", err)
+	}
 
-	if err := r.ReadMsg(&ch); err != nil {
-		if err == io.EOF {
+	// create chunk
+	addr := swarm.NewAddress(ch.Address)
+	chunk := swarm.NewChunk(addr, ch.Data)
+
+	// Select the closest peer to forward the chunk
+	peer, err := ps.peerSuggester.ClosestPeer(chunk.Address())
+	if err != nil {
+		// If i am the closest peer then store the chunk and send receipt
+		if errors.Is(err, topology.ErrWantSelf) {
+
+			// Store the chunk in the local store
+			_, err := ps.storer.Put(ctx, storage.ModePutSync, chunk)
+			if err != nil {
+				return fmt.Errorf("chunk store: %w", err)
+			}
+			ps.metrics.TotalChunksStoredInDB.Inc()
+
+			// Send a receipt immediately once the storage of the chunk is successfull
+			err = ps.sendReceipt(w, chunk.Address())
+			if err != nil {
+				return fmt.Errorf("send receipt: %w", err)
+			}
 			return nil
 		}
 		return err
 	}
 
-	// create chunk and store it in the local store
-	addr := swarm.NewAddress(ch.Address)
-	chunk := swarm.NewChunk(addr, ch.Data)
-	_, err := ps.storer.Put(ctx, storage.ModePutSync, chunk)
-	if err != nil {
+	//Dont forward this chunk to the same peer as the received peer
+	if p.Address.Equal(peer) {
 		return err
 	}
```

Comment:
This code block is returning the error which is nil.

This particular case is also an error which should be returned. If this happens, it means that a running node and a peer that is sending the chunk do not agree which one is closes to the node, as both are expecting that the another is closer.

This may need to be discussed, as it can happen and for now I think that it is ok to just return the concrete error, not nil error like currently.

Alternatively, this node can accept to store the chunk even if it wants to forward it back backward.

WDTY? @jmozah @acud

Reply:
I think we should store it and send a receipt back...